lib/healthcheck/worker.go

```diff
@@ -0,0 +1,441 @@
+	}
+	go w.run(ctx)
+	return w, nil
+}
+
+// worker perform health checks against a target resource and keeps track of
+// the target resource's health.
+type worker struct {
+	// cancel stops the worker permanently when called.
+	cancel context.CancelFunc
+	// clock is used to control time in tests.
+	clock clockwork.Clock
+	// healthCheckCfg is the current config used for performing health checks.
+	// Nil if no health check config currently matches the target resource.
+	healthCheckCfg *healthCheckConfig
+	// healthCheckConfigUpdateCh is used to notify the worker of a new health
+	// check config.
+	healthCheckConfigUpdateCh chan *healthCheckConfig
+	// healthCheckInterval is the current interval between health checks. Nil
+	// if there isn't currently a matching health check config for this worker.
+	healthCheckInterval *interval.Interval
+	// log emits logs.
+	log *slog.Logger
+	// dialFn is used to override the dialer in tests.
+	dialFn dialFunc
+
+	// lastResultErr is the last health check error or nil if there was no error.
+	lastResultErr error
+	// lastResultCount is the count of consecutive passing or failing health
+	// check results.
+	lastResultCount uint32
+	// lastResolvedEndpoints are the endpoints last resolved for a health check.
+	lastResolvedEndpoints []string
+
+	// mu guards concurrent access to the target resource and its health.
+	mu sync.RWMutex
+	// targetHealth is the latest target health. Initialized to "unknown" status
+	// before the worker starts.
+	targetHealth types.TargetHealth
+	// target is the health check target.
+	target Target
+
+	// healthUpdateCh is used to send/receive each target health update - only
+	// used in tests.
+	healthUpdateCh chan types.TargetHealth
+}
+
+// dialFunc dials an address on the given network.
+type dialFunc func(ctx context.Context, network, addr string) (net.Conn, error)
+
+// GetTargetHealth returns the worker's target health.
+func (w *worker) GetTargetHealth() *types.TargetHealth {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	return utils.CloneProtoMsg(&w.targetHealth)
+}
+
+// GetTargetResource returns the target resource.
+func (w *worker) GetTargetResource() types.ResourceWithLabels {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	return w.target.GetUpdatedResourceFn()
+}
+
+// UpdateHealthCheckConfig updates the worker to use a new health check
+// config.
+func (w *worker) UpdateHealthCheckConfig(newCfg *healthCheckConfig) {
+	// drain pending config update, if any
+	select {
+	case <-w.healthCheckConfigUpdateCh:
+	default:
+	}
+	w.healthCheckConfigUpdateCh <- newCfg
+}
+
+// Close closes the health checker.
+func (w *worker) Close() error {
+	w.cancel()
+	return nil
+}
+
+func (w *worker) run(ctx context.Context) {
+	defer func() {
+		if w.healthCheckInterval != nil {
+			w.healthCheckInterval.Stop()
+		}
+	}()
+
+	if w.healthCheckCfg != nil {
+		w.startHealthCheckInterval(ctx)
+		// no delay for the first health check after a target is registered
+		w.checkHealth(ctx)
```

Comment:
should the interval starts after the check? or maybe just set FirstDuration to 0 in the new interval? we can add a flag to `startHealthCheckInterval` to differentiate scenarios for first duration if needed.